ui/gtk4.c

```diff
@@ -0,0 +1,2176 @@
+             */
+            s->last_set = FALSE;
+            return ;
+        }
+    }
+    return ;
+}
+
+static void gd_button_event(GtkEventController *controller, gint n_presses,
+                            double x, double y, VirtualConsole *vc)
+{
+    GdkEvent *button = gtk_event_controller_get_current_event(controller);
+    GtkDisplayState *s = vc->s;
+    guint button_nbr = gdk_button_event_get_button(GDK_EVENT(button));
+    InputButton btn;
+
+    /* implicitly grab the input at the first click in the relative mode */
+    if (button_nbr == 1 && n_presses == 1 &&
+        !qemu_input_is_absolute() && s->ptr_owner != vc) {
+        if (!vc->window) {
+            s_action_set_state(s->grab_action, TRUE);
+        } else {
+            gd_grab_pointer(vc, "relative-mode-click");
+        }
+    }
+
+    if (button_nbr == 1) {
+        btn = INPUT_BUTTON_LEFT;
+    } else if (button_nbr == 2) {
+        btn = INPUT_BUTTON_MIDDLE;
+    } else if (button_nbr == 3) {
+        btn = INPUT_BUTTON_RIGHT;
+    } else if (button_nbr == 8) {
+        btn = INPUT_BUTTON_SIDE;
+    } else if (button_nbr == 9) {
+        btn = INPUT_BUTTON_EXTRA;
+    } else {
+        return;
+    }
+
+    if (n_presses == 2 || n_presses == 3) {
+        return;
+    }
+
+    qemu_input_queue_btn(vc->gfx.dcl.con, btn, n_presses == 1);
+    qemu_input_event_sync();
+}
+
+static gboolean gd_scroll_event(GtkEventController *controller,
+                                double delta_x, double delta_y,
+                                VirtualConsole *vc)
+{
+    GdkEvent *event = gtk_event_controller_get_current_event(controller);
+    GdkScrollDirection direction = gdk_scroll_event_get_direction(event);
+    InputButton btn_vertical;
+    InputButton btn_horizontal;
+    bool has_vertical = false;
+    bool has_horizontal = false;
+
+    if (direction == GDK_SCROLL_UP) {
+        btn_vertical = INPUT_BUTTON_WHEEL_UP;
+        has_vertical = true;
+    } else if (direction == GDK_SCROLL_DOWN) {
+        btn_vertical = INPUT_BUTTON_WHEEL_DOWN;
+        has_vertical = true;
+    } else if (direction == GDK_SCROLL_LEFT) {
+        btn_horizontal = INPUT_BUTTON_WHEEL_LEFT;
+        has_horizontal = true;
+    } else if (direction == GDK_SCROLL_RIGHT) {
+        btn_horizontal = INPUT_BUTTON_WHEEL_RIGHT;
+        has_horizontal = true;
+    } else if (direction == GDK_SCROLL_SMOOTH) {
+        if (delta_y > 0) {
+            btn_vertical = INPUT_BUTTON_WHEEL_DOWN;
+            has_vertical = true;
+        } else if (delta_y < 0) {
+            btn_vertical = INPUT_BUTTON_WHEEL_UP;
+            has_vertical = true;
+        } else if (delta_x > 0) {
+            btn_horizontal = INPUT_BUTTON_WHEEL_RIGHT;
+            has_horizontal = true;
+        } else if (delta_x < 0) {
+            btn_horizontal = INPUT_BUTTON_WHEEL_LEFT;
+            has_horizontal = true;
+        } else {
+            return TRUE;
+        }
+    } else {
+        return TRUE;
+    }
```

Comment:
Another `switch` statement here? And what does the return signifies? Isn't it always `TRUE`?

Reply:
The returns were not needed indeed and the switch is clearer for handling the various directions. Thanks!